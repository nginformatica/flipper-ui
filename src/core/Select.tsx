import {
    IconButton,
    InputAdornment,
    Select as MuiSelect
} from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'
import { Clear } from '@material-ui/icons'
import React, { ChangeEvent } from 'react'
import { DefaultProps } from './types'

interface SelectProps extends DefaultProps {
    autoWidth?: boolean
    value?: string | number
    multiple?: boolean
    variant?: 'standard' | 'outlined' | 'filled'
    hasClear?: boolean
    children: React.ReactNode
    onClear?: () => void
    onClose?: () => void
    onChange?: (
        event: ChangeEvent<HTMLSelectElement | HTMLInputElement>
    ) => void
}

const iconStyle = {
    right: '2px'
}

const renderEndAdornment = (onClear?: () => void) => (
    <InputAdornment position='end'>
        <IconButton onClick={onClear} size='small'>
            <Clear style={{ fontSize: '15px' }} />
        </IconButton>
    </InputAdornment>
)

const Select = ({
    children,
    style = {},
    margin,
    padding,
    hasClear,
    onClear,
    variant = 'outlined',
    ...otherProps
}: SelectProps) => {
    const useStyles = makeStyles(() => ({
        root: {
            padding: `10px 24px 10px ${hasClear ? '20' : '12'}px`
        },
        iconOutlined: hasClear
            ? { ...iconStyle, position: 'relative', marginLeft: '-20px' }
            : iconStyle
    }))
    const classes = useStyles()

    const hasValue = !!otherProps.value

    const endAdornment =
        hasValue && hasClear
            ? { endAdornment: renderEndAdornment(onClear) }
            : {}

    return (
        <MuiSelect
            {...endAdornment}
            variant={variant as 'outlined'}
            classes={{
                root: classes.root,
                icon: classes.iconOutlined
            }}
            style={{ margin, padding, ...style }}
            {...otherProps}>
            {children}
        </MuiSelect>
    )
}

export default Select
