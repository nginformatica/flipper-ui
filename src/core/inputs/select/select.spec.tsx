import React, { act } from 'react'
import { render, screen, waitFor } from '@testing-library/react'
import { userEvent } from '@testing-library/user-event'
import Select from '@/test/mocks/select-mock'
import '@testing-library/jest-dom'

describe('Select', () => {
    it('should render', async () => {
        render(<Select initialValue='0' />)

        const container = screen.getByText('Option 0')

        await act(async () => await userEvent.click(container))

        const options = screen.getAllByRole('option')

        expect(options.length).toBe(3)
    })

    it('should update value', async () => {
        render(<Select initialValue='0' />)

        const container = screen.getByText('Option 0')

        await act(async () => await userEvent.click(container))

        const option = screen.getByText('Option 1')

        await act(async () => await userEvent.click(option))

        expect((container as HTMLDivElement).innerHTML).toBe('Option 1')
    })

    it('should call onChange with correct values', async () => {
        const onChangeSpy = jest.fn()

        render(
            <Select initialValue='0' selectProps={{ onChange: onChangeSpy }} />
        )

        const container = screen.getByText('Option 0')

        await act(async () => await userEvent.click(container))

        const option = screen.getByText('Option 1')

        await act(async () => await userEvent.click(option))

        waitFor(() => {
            const input = screen.getByTestId('select-input')

            expect(input).toHaveValue('1')
        })
    })

    it('should render with clear button', async () => {
        render(<Select initialValue='1' selectProps={{ hasClear: true }} />)

        screen.getByText('Option 1')

        const clearButton = screen.getByRole('end-adornment-component')

        await act(async () => await userEvent.click(clearButton))

        const input = screen.getByTestId('select-input') as HTMLInputElement

        expect(input.value).toBe('')
    })

    it('should match snapshot', () => {
        const { container } = render(
            <Select initialValue='1' selectProps={{ onChange: jest.fn() }} />
        )

        expect(container).toMatchSnapshot()
    })
})
