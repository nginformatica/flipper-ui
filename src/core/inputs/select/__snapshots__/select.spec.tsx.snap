// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should match snapshot 1`] = `
<div>
  <div
    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-adornedEnd  css-2d5w1o-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
  >
    <div
      aria-controls=":r5:"
      aria-expanded="false"
      aria-haspopup="listbox"
      class="MuiSelect-select makeStyles-select-11 MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-4ygxeq-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
      role="combobox"
      tabindex="0"
    >
      <span
        class="notranslate"
      >
        ​
      </span>
    </div>
    <input
      aria-hidden="true"
      aria-invalid="false"
      class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
      data-testid="select-input"
      tabindex="-1"
      value="1"
    />
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon makeStyles-iconOutlined-12 MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
      data-testid="ArrowDropDownIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
    </svg>
    <div
      class="MuiInputAdornment-root MuiInputAdornment-positionEnd css-1laqsz7-MuiInputAdornment-root"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
        role="end-adornment-component"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ClearIcon"
          focusable="false"
          style="font-size: 15px;"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <fieldset
      aria-hidden="true"
      class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        class="css-ihdtdm"
      >
        <span
          class="notranslate"
        >
          ​
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;
